aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,948 @@
+                                } else {
+                                    // No action required, the key is not due for expiration.
+                                    None
+                                }
+                            } else {
+                                info!(
+                                    ?ciphersuite,
+                                    id = ?keypair.id(),
+                                    "expiring extraneous active key"
+                                );
+                                Some(Op::Update(*keypair.id(), HpkeKeyState::Expired))
+                            }
+                        } else {
+                            info!(
+                                ciphersuite = ?keypair.ciphersuite(),
+                                id = ?keypair.id(),
+                                "expiring active key for unknown ciphersuite"
+                            );
+                            Some(Op::Update(id, HpkeKeyState::Expired))
+                        }
+                    }
+                    HpkeKeyState::Pending => {
+                        if ciphersuite_known {
+                            let ciphersuite = keypair.ciphersuite();
+                            let latest_pending_keypair = self
+                                .keypairs
+                                .values()
+                                .filter(|candidate| {
+                                    keypair.ciphersuite() == candidate.ciphersuite()
+                                        && keypair.is_pending()
+                                })
+                                .max_by_key(|keypair| keypair.last_state_change_at());
+
+                            if Some(keypair) == latest_pending_keypair {
+                                if duration_since(&self.clock, keypair.last_state_change_at())
+                                    > self.config.pending_duration
+                                {
+                                    info!(
+                                        id = ?keypair.id(),
+                                        ?ciphersuite,
+                                        "promoting pending keypair to active"
+                                    );
+                                    Some(Op::Update(*keypair.id(), HpkeKeyState::Active))
+                                } else {
+                                    // No action required, the key is pending but it's not ready
+                                    // for promotion.
+                                    None
+                                }
+                            } else {
+                                info!(
+                                    id = ?keypair.id(),
+                                    ?ciphersuite,
+                                    "deleting extraneous pending keypair"
+                                );
+                                Some(Op::Delete(id))
+                            }
```

Comment:
The following pathological case occurs to me. Suppose we have, all with the same ciphersuite, one active keypair that is old enough to be expired, and two pending keypairs, one that is old enough to be promoted, and one that is not. The logic under the `Active` arm will see that there's a pending keypair that's ready, and thus expire the active one. The logic under the `Pending` arm, however, will not promote either pending keypair to active. The older pending keypair, which is ready for promotion, will be deleted because it is not the latest pending keypair. The newer pending keypair, which is not ready for promotion, will have no action taken on it.

I think this bug arose when transforming the logic to consider one keypair at a time. Ideally the expiration and promotion should either both happen or not, but considering them separately created a hole because the preconditions weren't quite the same.

There are a few ways we could fix this: 
- Changing the criteria for which pending keypair we delete
- Writing more complicated conditions for promotion and expiration, so that either both happens or neither happens
- Making processing two-pass, so that we can choose whether to expire an active key based on whether we actually have promoted a pending key
- Replacing the `keypairs.iter.filter_map()` loop, and pushing multiple operations onto the `Vec` together, so we don't get out of sync.

We could also improve tests to catch this sort of thing by using `quickcheck` or similar to generate the database's initial state and the key rotator's configuration, then asserting various correctness properties about the initial and final states.

Reply:
Nice find. That lends a lot of credence to @branlwyd point about precondition mismatch being a possible source of bugs.

Reply:
I've solved this by more or less adopting @branlwyd's strategy of handling each ciphersuite independently. This allows us to make the state changes in lockstep and based on the same preconditions.

Reply:
I'm still looking into the testing strategy you suggested, but the logic is ready for another pass.